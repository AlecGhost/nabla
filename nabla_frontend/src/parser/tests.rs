use crate::{
    ast::*,
    lexer::lex,
    parser::{parse, Error, ErrorMessage},
    token::{Token, TokenRange, TokenType},
};
use pretty_assertions::assert_eq;

fn ident(name: &str, range: TokenRange) -> Ident {
    Ident {
        name: name.to_string(),
        info: AstInfo::new(range),
    }
}

fn info(range: TokenRange) -> AstInfo {
    AstInfo::new(range)
}

#[test]
fn empty() {
    let src = "";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: Vec::new(),
            info: AstInfo::new(0..1),
        },
        program
    );
}

#[test]
fn token_after_eof() {
    let tokens = vec![
        Token::new(TokenType::Eof, 0..0),
        Token::new(TokenType::Eof, 0..0),
    ];
    let (_, errors) = parse(&tokens);
    assert_eq!(vec![Error::new(ErrorMessage::TokensAfterEof, 1..1)], errors);
}

#[test]
fn missing_type_expr() {
    let src = "let a: = {}";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (_, errors) = parse(&tokens);
    assert_eq!(vec![Error::new(ErrorMessage::ExpectedExpr, 3..3)], errors);
}

#[test]
fn missing_expr() {
    let src = "let a =";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (_, errors) = parse(&tokens);
    assert_eq!(vec![Error::new(ErrorMessage::ExpectedExpr, 4..4)], errors);
}

#[test]
fn missing_multiple_exprs() {
    let src = "let a: =";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (_, errors) = parse(&tokens);
    assert_eq!(
        vec![
            Error::new(ErrorMessage::ExpectedExpr, 3..3),
            Error::new(ErrorMessage::ExpectedExpr, 5..5),
        ],
        errors
    );
}

#[test]
fn use_simple() {
    let src = "use a";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Use(Use {
                use_kw: info(0..1),
                name: Some(ident("a", 1..3)),
                body: None,
                info: info(0..3),
            })],
            info: info(0..4),
        },
        program
    );
}

#[test]
fn use_all() {
    let src = "use a::*";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Use(Use {
                use_kw: info(0..1),
                name: Some(ident("a", 1..3)),
                body: Some(UseBody {
                    double_colon: info(3..4),
                    kind: Some(UseKind::All(info(4..5))),
                    info: info(3..5),
                }),
                info: info(0..5),
            })],
            info: info(0..6),
        },
        program
    );
}

#[test]
fn use_single() {
    let src = "use a::b";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Use(Use {
                use_kw: info(0..1),
                name: Some(ident("a", 1..3)),
                body: Some(UseBody {
                    double_colon: info(3..4),
                    kind: Some(UseKind::Single(UseItem {
                        name: ident("b", 4..5),
                        body: None,
                        alias: None,
                        info: info(4..5),
                    })),
                    info: info(3..5),
                }),
                info: info(0..5),
            })],
            info: info(0..6),
        },
        program
    );
}

#[test]
fn use_multiple() {
    let src = "use a::{b c}";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Use(Use {
                use_kw: info(0..1),
                name: Some(ident("a", 1..3)),
                body: Some(UseBody {
                    double_colon: info(3..4),
                    kind: Some(UseKind::Multiple(UseItems {
                        lcurly: info(4..5),
                        items: vec![
                            UseItem {
                                name: ident("b", 5..6),
                                body: None,
                                alias: None,
                                info: info(5..6),
                            },
                            UseItem {
                                name: ident("c", 6..8),
                                body: None,
                                alias: None,
                                info: info(6..8),
                            },
                        ],
                        rcurly: Some(info(8..9)),
                        info: info(4..9),
                    })),
                    info: info(3..9),
                }),
                info: info(0..9),
            })],
            info: info(0..10),
        },
        program
    );
}

#[test]
fn use_complex() {
    let src = "use a::{b::{ c::d as x e::* } f as y}";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    insta::assert_debug_snapshot!(program);
}

#[test]
fn def_ident() {
    let src = "def x = y";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Def(Def {
                def_kw: info(0..1),
                name: Some(ident("x", 1..3)),
                colon: None,
                type_expr: None,
                eq: Some(info(3..5)),
                expr: Some(Expr::Single(Single::Named(Named {
                    name: ident("y", 5..7),
                    inner_names: Vec::new(),
                    expr: None,
                    info: info(5..7),
                }))),
                info: info(0..7),
            })],
            info: info(0..8),
        },
        program
    );
}

#[test]
fn def_union() {
    let src = "def ok = \"yes\" | true";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Def(Def {
                def_kw: info(0..1),
                name: Some(ident("ok", 1..3)),
                colon: None,
                type_expr: None,
                eq: Some(info(3..5)),
                expr: Some(Expr::Union(Union {
                    single: Single::Primitive(Primitive::String(PrimitiveValue {
                        value: "yes".to_string(),
                        info: info(5..7),
                    })),
                    alternatives: vec![UnionAlternative {
                        pipe: info(7..9),
                        single: Some(Single::Primitive(Primitive::Bool(Bool::new_true(info(
                            9..11
                        ))))),
                        info: info(7..11),
                    }],
                    info: info(5..11),
                })),
                info: info(0..11),
            })],
            info: info(0..12),
        },
        program
    );
}

#[test]
fn def_struct() {
    let src = "
def Person = {
    name: string
    age: number = 0
}";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Def(Def {
                def_kw: info(0..2),
                name: Some(ident("Person", 2..4)),
                colon: None,
                type_expr: None,
                eq: Some(info(4..6)),
                expr: Some(Expr::Single(Single::Struct(Struct {
                    lcurly: info(6..8),
                    fields: vec![
                        StructField {
                            name: ident("name", 8..10),
                            colon: Some(info(10..11)),
                            type_expr: Some(Expr::Single(Single::Named(Named {
                                name: ident("string", 11..13),
                                inner_names: Vec::new(),
                                expr: None,
                                info: info(11..13),
                            }))),
                            eq: None,
                            expr: None,
                            alias: None,
                            info: info(8..13),
                        },
                        StructField {
                            name: ident("age", 13..15),
                            colon: Some(info(15..16)),
                            type_expr: Some(Expr::Single(Single::Named(Named {
                                name: ident("number", 16..18),
                                inner_names: Vec::new(),
                                expr: None,
                                info: info(16..18),
                            }))),
                            eq: Some(info(18..20)),
                            expr: Some(Expr::Single(Single::Primitive(Primitive::Number(
                                PrimitiveValue {
                                    value: "0".to_string(),
                                    info: info(20..22),
                                }
                            )))),
                            alias: None,
                            info: info(13..22),
                        },
                    ],
                    rcurly: Some(info(22..24)),
                    info: info(6..24),
                }))),
                info: info(0..24),
            })],
            info: info(0..25),
        },
        program
    );
}

#[test]
fn def_list() {
    let src = "def Strings = [ string ]";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    assert_eq!(
        Program {
            globals: vec![Global::Def(Def {
                def_kw: info(0..1),
                name: Some(ident("Strings", 1..3)),
                colon: None,
                type_expr: None,
                eq: Some(info(3..5)),
                expr: Some(Expr::Single(Single::List(List {
                    lbracket: info(5..7),
                    exprs: vec![Expr::Single(Single::Named(Named {
                        name: ident("string", 7..9),
                        inner_names: Vec::new(),
                        expr: None,
                        info: info(7..9),
                    }))],
                    rbracket: Some(info(9..11)),
                    info: info(5..11),
                }))),
                info: info(0..11),
            })],
            info: info(0..12),
        },
        program
    );
}

#[test]
fn def_all_syntax() {
    let src = "
use other_dir::*
use dir2::{x::{y} z}
use dir3::test

def x = {
    only_type: string
    only_expr = \"expr\"
    type_and_expr: number = 1
    only_type_as: string as \"x\"
    only_expr_as = \"expr\" as \"y\"
    type_and_expr_as: number = 1 as \"z\"
}
";
    let (tokens, errors) = lex(src);
    assert!(errors.is_empty());
    let (program, errors) = parse(&tokens);
    assert!(errors.is_empty());
    insta::assert_debug_snapshot!(program);
}
