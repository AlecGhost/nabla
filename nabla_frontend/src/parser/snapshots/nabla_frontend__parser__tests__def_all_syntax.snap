---
source: nabla_frontend/src/parser/tests.rs
expression: program
---
Program {
    globals: [
        Use(
            Use {
                use_kw: AstInfo {
                    range: 0..2,
                },
                name: Some(
                    Ident {
                        name: "other_dir",
                        info: AstInfo {
                            range: 2..4,
                        },
                    },
                ),
                body: Some(
                    UseBody {
                        double_colon: AstInfo {
                            range: 4..5,
                        },
                        kind: Some(
                            All(
                                AstInfo {
                                    range: 5..6,
                                },
                            ),
                        ),
                        info: AstInfo {
                            range: 4..6,
                        },
                    },
                ),
                info: AstInfo {
                    range: 0..6,
                },
            },
        ),
        Use(
            Use {
                use_kw: AstInfo {
                    range: 6..8,
                },
                name: Some(
                    Ident {
                        name: "dir2",
                        info: AstInfo {
                            range: 8..10,
                        },
                    },
                ),
                body: Some(
                    UseBody {
                        double_colon: AstInfo {
                            range: 10..11,
                        },
                        kind: Some(
                            Multiple(
                                UseItems {
                                    lcurly: AstInfo {
                                        range: 11..12,
                                    },
                                    items: [
                                        UseItem {
                                            name: Ident {
                                                name: "x",
                                                info: AstInfo {
                                                    range: 12..13,
                                                },
                                            },
                                            body: Some(
                                                UseBody {
                                                    double_colon: AstInfo {
                                                        range: 13..14,
                                                    },
                                                    kind: Some(
                                                        Multiple(
                                                            UseItems {
                                                                lcurly: AstInfo {
                                                                    range: 14..15,
                                                                },
                                                                items: [
                                                                    UseItem {
                                                                        name: Ident {
                                                                            name: "y",
                                                                            info: AstInfo {
                                                                                range: 15..16,
                                                                            },
                                                                        },
                                                                        body: None,
                                                                        alias: None,
                                                                        info: AstInfo {
                                                                            range: 15..16,
                                                                        },
                                                                    },
                                                                ],
                                                                rcurly: Some(
                                                                    AstInfo {
                                                                        range: 16..17,
                                                                    },
                                                                ),
                                                                info: AstInfo {
                                                                    range: 14..17,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    info: AstInfo {
                                                        range: 13..17,
                                                    },
                                                },
                                            ),
                                            alias: None,
                                            info: AstInfo {
                                                range: 12..17,
                                            },
                                        },
                                        UseItem {
                                            name: Ident {
                                                name: "z",
                                                info: AstInfo {
                                                    range: 17..19,
                                                },
                                            },
                                            body: None,
                                            alias: None,
                                            info: AstInfo {
                                                range: 17..19,
                                            },
                                        },
                                    ],
                                    rcurly: Some(
                                        AstInfo {
                                            range: 19..20,
                                        },
                                    ),
                                    info: AstInfo {
                                        range: 11..20,
                                    },
                                },
                            ),
                        ),
                        info: AstInfo {
                            range: 10..20,
                        },
                    },
                ),
                info: AstInfo {
                    range: 6..20,
                },
            },
        ),
        Use(
            Use {
                use_kw: AstInfo {
                    range: 20..22,
                },
                name: Some(
                    Ident {
                        name: "dir3",
                        info: AstInfo {
                            range: 22..24,
                        },
                    },
                ),
                body: Some(
                    UseBody {
                        double_colon: AstInfo {
                            range: 24..25,
                        },
                        kind: Some(
                            Single(
                                UseItem {
                                    name: Ident {
                                        name: "test",
                                        info: AstInfo {
                                            range: 25..26,
                                        },
                                    },
                                    body: None,
                                    alias: None,
                                    info: AstInfo {
                                        range: 25..26,
                                    },
                                },
                            ),
                        ),
                        info: AstInfo {
                            range: 24..26,
                        },
                    },
                ),
                info: AstInfo {
                    range: 20..26,
                },
            },
        ),
        Def(
            Def {
                def_kw: AstInfo {
                    range: 26..28,
                },
                name: Some(
                    Ident {
                        name: "x",
                        info: AstInfo {
                            range: 28..30,
                        },
                    },
                ),
                colon: None,
                type_expr: None,
                eq: Some(
                    AstInfo {
                        range: 30..32,
                    },
                ),
                expr: Some(
                    Single(
                        Struct(
                            Struct {
                                lcurly: AstInfo {
                                    range: 32..34,
                                },
                                fields: [
                                    StructField {
                                        name: Ident {
                                            name: "only_type",
                                            info: AstInfo {
                                                range: 34..36,
                                            },
                                        },
                                        colon: Some(
                                            AstInfo {
                                                range: 36..37,
                                            },
                                        ),
                                        type_expr: Some(
                                            Single(
                                                Named(
                                                    Named {
                                                        name: Ident {
                                                            name: "string",
                                                            info: AstInfo {
                                                                range: 37..39,
                                                            },
                                                        },
                                                        inner_names: [],
                                                        expr: None,
                                                        info: AstInfo {
                                                            range: 37..39,
                                                        },
                                                    },
                                                ),
                                            ),
                                        ),
                                        eq: None,
                                        expr: None,
                                        alias: None,
                                        info: AstInfo {
                                            range: 34..39,
                                        },
                                    },
                                    StructField {
                                        name: Ident {
                                            name: "only_expr",
                                            info: AstInfo {
                                                range: 39..41,
                                            },
                                        },
                                        colon: None,
                                        type_expr: None,
                                        eq: Some(
                                            AstInfo {
                                                range: 41..43,
                                            },
                                        ),
                                        expr: Some(
                                            Single(
                                                Primitive(
                                                    String(
                                                        PrimitiveValue {
                                                            value: "expr",
                                                            info: AstInfo {
                                                                range: 43..45,
                                                            },
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ),
                                        alias: None,
                                        info: AstInfo {
                                            range: 39..45,
                                        },
                                    },
                                    StructField {
                                        name: Ident {
                                            name: "type_and_expr",
                                            info: AstInfo {
                                                range: 45..47,
                                            },
                                        },
                                        colon: Some(
                                            AstInfo {
                                                range: 47..48,
                                            },
                                        ),
                                        type_expr: Some(
                                            Single(
                                                Named(
                                                    Named {
                                                        name: Ident {
                                                            name: "number",
                                                            info: AstInfo {
                                                                range: 48..50,
                                                            },
                                                        },
                                                        inner_names: [],
                                                        expr: None,
                                                        info: AstInfo {
                                                            range: 48..50,
                                                        },
                                                    },
                                                ),
                                            ),
                                        ),
                                        eq: Some(
                                            AstInfo {
                                                range: 50..52,
                                            },
                                        ),
                                        expr: Some(
                                            Single(
                                                Primitive(
                                                    Number(
                                                        PrimitiveValue {
                                                            value: "1",
                                                            info: AstInfo {
                                                                range: 52..54,
                                                            },
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ),
                                        alias: None,
                                        info: AstInfo {
                                            range: 45..54,
                                        },
                                    },
                                    StructField {
                                        name: Ident {
                                            name: "only_type_as",
                                            info: AstInfo {
                                                range: 54..56,
                                            },
                                        },
                                        colon: Some(
                                            AstInfo {
                                                range: 56..57,
                                            },
                                        ),
                                        type_expr: Some(
                                            Single(
                                                Named(
                                                    Named {
                                                        name: Ident {
                                                            name: "string",
                                                            info: AstInfo {
                                                                range: 57..59,
                                                            },
                                                        },
                                                        inner_names: [],
                                                        expr: None,
                                                        info: AstInfo {
                                                            range: 57..59,
                                                        },
                                                    },
                                                ),
                                            ),
                                        ),
                                        eq: None,
                                        expr: None,
                                        alias: Some(
                                            Alias {
                                                as_kw: AstInfo {
                                                    range: 59..61,
                                                },
                                                name: Some(
                                                    String(
                                                        PrimitiveValue {
                                                            value: "x",
                                                            info: AstInfo {
                                                                range: 61..63,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                info: AstInfo {
                                                    range: 59..63,
                                                },
                                            },
                                        ),
                                        info: AstInfo {
                                            range: 54..63,
                                        },
                                    },
                                    StructField {
                                        name: Ident {
                                            name: "only_expr_as",
                                            info: AstInfo {
                                                range: 63..65,
                                            },
                                        },
                                        colon: None,
                                        type_expr: None,
                                        eq: Some(
                                            AstInfo {
                                                range: 65..67,
                                            },
                                        ),
                                        expr: Some(
                                            Single(
                                                Primitive(
                                                    String(
                                                        PrimitiveValue {
                                                            value: "expr",
                                                            info: AstInfo {
                                                                range: 67..69,
                                                            },
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ),
                                        alias: Some(
                                            Alias {
                                                as_kw: AstInfo {
                                                    range: 69..71,
                                                },
                                                name: Some(
                                                    String(
                                                        PrimitiveValue {
                                                            value: "y",
                                                            info: AstInfo {
                                                                range: 71..73,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                info: AstInfo {
                                                    range: 69..73,
                                                },
                                            },
                                        ),
                                        info: AstInfo {
                                            range: 63..73,
                                        },
                                    },
                                    StructField {
                                        name: Ident {
                                            name: "type_and_expr_as",
                                            info: AstInfo {
                                                range: 73..75,
                                            },
                                        },
                                        colon: Some(
                                            AstInfo {
                                                range: 75..76,
                                            },
                                        ),
                                        type_expr: Some(
                                            Single(
                                                Named(
                                                    Named {
                                                        name: Ident {
                                                            name: "number",
                                                            info: AstInfo {
                                                                range: 76..78,
                                                            },
                                                        },
                                                        inner_names: [],
                                                        expr: None,
                                                        info: AstInfo {
                                                            range: 76..78,
                                                        },
                                                    },
                                                ),
                                            ),
                                        ),
                                        eq: Some(
                                            AstInfo {
                                                range: 78..80,
                                            },
                                        ),
                                        expr: Some(
                                            Single(
                                                Primitive(
                                                    Number(
                                                        PrimitiveValue {
                                                            value: "1",
                                                            info: AstInfo {
                                                                range: 80..82,
                                                            },
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ),
                                        alias: Some(
                                            Alias {
                                                as_kw: AstInfo {
                                                    range: 82..84,
                                                },
                                                name: Some(
                                                    String(
                                                        PrimitiveValue {
                                                            value: "z",
                                                            info: AstInfo {
                                                                range: 84..86,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                info: AstInfo {
                                                    range: 82..86,
                                                },
                                            },
                                        ),
                                        info: AstInfo {
                                            range: 73..86,
                                        },
                                    },
                                ],
                                rcurly: Some(
                                    AstInfo {
                                        range: 86..88,
                                    },
                                ),
                                info: AstInfo {
                                    range: 32..88,
                                },
                            },
                        ),
                    ),
                ),
                info: AstInfo {
                    range: 26..88,
                },
            },
        ),
    ],
    info: AstInfo {
        range: 0..90,
    },
}
